File: src/site/eventim_light.rs
use std::borrow::Borrow;

use lazy_static::lazy_static;
use regex::Regex;
use serde_json::Value;

use crate::site::Filter;
use crate::tools::HTTP;
use crate::Event;

pub struct EventimLight {
    collected_events: Vec<Value>,
}

impl EventimLight {
    pub fn new(venue: String, http: &HTTP) -> Self {
        lazy_static! {
            static ref REG: Regex =
                Regex::new("(?i)window.__INITIAL_STATE__=(\\{.*?\\});").unwrap();
        }

        let mut collected_events: Vec<Value> = Vec::new();

        let html = http.get(("https://www.eventim-light.com/de/a/".to_owned() + &venue).as_str());

        let matches = REG.captures(html.borrow()).unwrap();
        let text: &str = matches.get(1).map_or("", |m| m.as_str());
        let json: Value = serde_json::from_str(text).unwrap();
        for (_, event) in json["events"]["cache"].as_object().unwrap() {
            collected_events.push(event.to_owned());
        }

        Self { collected_events }
    }
}

impl Filter for EventimLight {
    fn is_it_metal(&self, evt: &Event) -> bool {
        let date = evt.date.format("%Y-%m-%d").to_string();
        let date = date.as_str();
        for collected_event in self.collected_events.iter() {
            if collected_event["category"].as_str().unwrap() == "Hard & Heavy"
                && collected_event["start"].to_string().starts_with(date)
            {
                return true;
            }
        }

        false
    }
}

File: src/Site/EventimLight.php
<?php

declare(strict_types=1);

namespace MetalLE\Site;

class EventimLight
{
    public function fetchDates(string $venue): iterable
    {
        $customHeaders =
            "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:96.0) Gecko/20100101 Firefox/96.0\r\n"
            ."Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8\r\n"
            ."Accept-Language: de,en-US;q=0.7,en;q=0.3\r\n"
            ."Accept-Encoding: identity\r\n"
            ."DNT: 1\r\n"
            ."Connection: close\r\n"
            ."Pragma: no-cache\r\n"
            ."Cache-Control: no-cache\r\n"
        ;
        $context = stream_context_create(['http' => ['method' => 'GET', 'header' => $customHeaders]]);

        $plainHtml = file_get_contents(
            'https://www.eventim-light.com/de/a/' . $venue,
            false,
            $context,
        );
        if (preg_match('#window.__INITIAL_STATE__=({.*});#iU', $plainHtml, $matches)) {
            $data = json_decode($matches[1], false, 512, JSON_THROW_ON_ERROR);
            foreach ($data->events->cache as $item) {
                if ($item->category === 'Hard & Heavy') {
                    yield $item->title => new \DateTimeImmutable($item->start);
                }
            }
        }

        return [];
    }
}